lightly/api/api_workflow_upload_metadata.py

```diff
@@ -168,48 +142,62 @@ def upload_custom_metadata(self,
 
         self.verify_custom_metadata_format(custom_metadata)
 
-        # create a mapping from sample filenames to custom metadata
+
+
+        # For each metadata, we need the corresponding sample_id
+        # on the server. The mapping is:
+        # metadata -> image_id -> filename -> sample_id
+
+        image_id_to_filename = {
+            image_info[COCO_ANNOTATION_KEYS.images_id]:
+                image_info[COCO_ANNOTATION_KEYS.images_filename]
+            for image_info in custom_metadata[COCO_ANNOTATION_KEYS.images]
+        }
+
         samples = self._samples_api.get_samples_by_dataset_id(self.dataset_id)
-        filename_to_metadata = self.index_custom_metadata_by_filename(
-            [sample.file_name for sample in samples],
-            custom_metadata,
-        )
-        if len(filename_to_metadata) != len(custom_metadata[COCO_ANNOTATION_KEYS.images]):
-            raise ValueError(
-                f'There is a mismatch between the number of images '
-                f'in the metadata file ({len(filename_to_metadata)}) and on the '
-                f'server ({len(custom_metadata[COCO_ANNOTATION_KEYS.images])}).'
-            )
+        filename_to_sample_id = {
+            sample.file_name: sample.id
+            for sample in samples
+        }
+
+        upload_requests = []
+        for metadata in custom_metadata[COCO_ANNOTATION_KEYS.custom_metadata]:
+            image_id = metadata[COCO_ANNOTATION_KEYS.custom_metadata_image_id]
+            filename = image_id_to_filename.get(image_id, None)
+            if filename is None:
+                raise ValueError(f"Your custom_metadata file is malformatted. "
+                                 f"There is an annotation with an image_id of "
+                                 f"{{{image_id}}},"
+                                 f"but the list of images does not include this id.")
+            sample_id = filename_to_sample_id.get(filename, None)
+            if sample_id is None:
+                raise ValueError(f"Your custom_metadata file "
```

Comment:
```python
InvalidCustomMetadataError(
    f' Sample with filename {filename} does not exist on the server'
)
```

Reply:
And tell the user that this annotations is skipped.

Reply:
This is how the implementation looks like:


![image](https://user-images.githubusercontent.com/20324507/157717173-bb66b902-b246-4f59-91e6-45b6bf8e8b02.jpeg)

Reply:
> Your proposals are a bit too short in my opinion. I would also try to hint the user more at how to solve the problem.
Fair point, maybe we can have a tech talk about how to write good exceptions sometime, I bet there are guidelines 🙂

Reply:
Only thing I could find so far:
https://google.github.io/styleguide/pyguide.html#3102-error-messages